Adding instructions for installing pillow (#312)
cheran-senthil/TLE: README.md

# TLE

TLE is a Discord bot centered around Competitive Programming.

## Features

The features of the bot are split into a number of cogs, each handling their own set of commands.

### Codeforces cogs

- **Codeforces** Commands that can recommend problems or contests to users, taking their rating into account.
- **Contests** Shows details of upcoming/running contests.
- **Graphs** Plots various data gathered from Codeforces, e.g. rating distributions and user problem statistics.
- **Handles** Gets or sets information about a specific user's Codeforces handle, or shows a list of Codeforces handles.

### CSES cog

- **CSES** Commands related to the [CSES problemset](https://cses.fi/problemset/), such as showing leaderboards.

### Other cogs

- **Starboard** Commands related to the starboard, which adds messages to a specific channel when enough users react with a ⭐️.
- **CacheControl** Commands related to data caching.

## Installation

Clone the repository

```bash
git clone https://github.com/cheran-senthil/TLE
```

> Now all dependencies need to be installed. TLE uses [Poetry](https://poetry.eustace.io/) to manage dependencies. After installing Poetry navigate to the root of the repo and run

Reviewer: ```suggestion
### Dependencies

Now all dependencies need to be installed. TLE uses [Poetry](https://poetry.eustace.io/) to manage its python dependencies. After installing Poetry navigate to the root of the repo and run
```

> :warning: **TLE requires Python 3.7 or later!**

If you are using Ubuntu with older versions of python, then do the following:

```bash
apt-get install python3.7-venv libpython3.7-dev
python3.7 -m pip install poetry
python3.7 -m poetry install
```

---

#### Library dependencies

TLE also depends on cairo and pango for graphics and text rendering, which you need to install. For Ubuntu, the relevant packages can be installed with:

```bash
apt-get install libcairo2-dev libgirepository1.0-dev libpango1.0-dev pkg-config python3-dev gir1.2-pango-1.0
```

Additionally TLE uses pillow for graphics, which requires the following packages:

```bash
apt-get install libjpeg-dev zlib1g-dev
```

### Final steps

You will need to setup a bot on your server before continuing, follow the directions [here](https://github.com/reactiflux/discord-irc/wiki/Creating-a-discord-bot-&-getting-a-token). Following this, you should have your bot appearing in your server and you should have the Discord bot token.

To start TLE export the token as an environment variable and run TLE using:

```bash

> export BOT_TOKEN="<BOT_TOKEN_FROM_DISCORD_CONSOLE>"

Author: ```suggestion
export BOT_TOKEN="XXXXXXXXXXXXXXXXXXXXXXXX.XXXXXX.XXXXXXXXXXXXXXXXXXXXXXXXXXX"
```

---

> TLE also depends on cairo and pango for graphics and text rendering, which you need to install through your package manager. For Ubuntu, the relevant packages can be installed with:

Reviewer: ```suggestion
#### Library dependencies

TLE also depends on cairo and pango for graphics and text rendering, which you need to install. For Ubuntu, the relevant packages can be installed with:
```

```bash
apt-get install libcairo2-dev libgirepository1.0-dev libpango1.0-dev pkg-config python3-dev gir1.2-pango-1.0
```

> TLE also depends on pillow, which needs to install the following packages:

Reviewer: ```suggestion
Additionally TLE uses pillow for graphics, which requires the following packages:
```

```bash
apt-get install libjpeg-dev zlib1g-dev
```

> ---

Reviewer: ```suggestion
### Final steps
```

You will need to setup a bot on your server before continuing, follow the directions [here](https://github.com/reactiflux/discord-irc/wiki/Creating-a-discord-bot-&-getting-a-token). Following this, you should have your bot appearing in your server and you should have the Discord bot token.

To start TLE export the token as an environment variable

```bash
export BOT_TOKEN="<BOT_TOKEN_FROM_DISCORD_CONSOLE>"
```

and run

```bash
poetry run python -m tle

> ```

Reviewer: Seems like github doesn't like suggestions that covers deleted lines. But we could make this simper. Additionally adding something about `run.sh` might make sense. Could leave that for later though.

---

    Export the token you find in your discord console as an environment variable and run TLE using
    
    ```bash
    export BOT_TOKEN="XXXXXXXXXXXXXXXXXXXXXXXX.XXXXXX.XXXXXXXXXXXXXXXXXXXXXXXXXXX"
    poetry run python -m tle
    ```

Author: I prefer leaving `run.sh` for later because I also want to replace `environment with `.env` as per the convention or at least adding `environment` to `.gitignore`.

Reviewer: Leaving it for later sounds good. I think .env is a python venv convention, so we don't want to conflict with that.

Reviewer: So, the reason I wanted to rephrase this part is to make the expected format of the token very obvious. I've had to help someone debug who happened to copy the wrong token.
